<a href="https://colab.research.google.com/github/dieguito1331/cienciaDatos/blob/master/posgrado/teoria/01 - Python basico.ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Introducción a Python ##

Python es un lenguaje interpretado de tipos dinámicos. A diferencia de R, por ejemplo, es un lenguaje de uso general y su capacidad para las ciencias de datos no viene dada desde sus librerías internas sino por los módulos externos. Se puede usar en forma imperativa u orientada a objetos; teóricamente también dentro del paradigma de programación funcional, pero con ciertas limitaciones (existen límites para ejecución recursiva, por ejemplo).

Desde lo estético, Python reemplaza las llaves por indentaciones, cosa que fuerza un código más leíble. Por lo demás, existe toda una "filosofía" de lo que es "pythonico", en términos de prolijidad, funcionalidad y belleza del código.

Otra característica de este lenguaje es que es un lenguaje Interpretado, esto significa que que no debe ser compilado. En resumidas cuentas, se cuenta con un intérprete que va a traduciendo el código escrito en un código máquina a medida que corre el mismo.

### Historia de Python ###
El inicio de este lenguaje se remonta en los años '90. Creado en el centro de investigación Holandés Centrum Wiskunde & Informatica (CWI) por Guido Van Rossum. Se basó en otro lenguaje de su autoría denominado "ABC" y la idea que implementó fue hacer un lenguaje sencillo de aprender y lo más parecido al lenguaje humano.


## Diversos Usos de Python ##

Cómo mencionamos anteriormente, Python es utilizado para una amplia gama de desarrollos (Desde desarrollos web, hasta Machine Learning, pasando por desarrollos de Inteligencia Artificial, etc)

Python sumado a las tecnologías que nos entregan los diferentes proveedores en la nube, se pueden realizar implementaciones mucho más facil que con tecnologías más antiguas.




In [ ]:
# Manejo de cadenas
string = "esto es una cadena"
print(string)

esto es una cadena


In [ ]:
import os
os.getcwd()

'D:\\Users\\900214017\\Documents\\proyectos\\cursoIUDU\\cursoiudu'

In [ ]:
type(string)

str

In [ ]:
string = 5
print(string)

5


In [ ]:
type(string)

int

In [ ]:
string_1 = "esto es "
string_2 = "una cadena"
print(string_1+string_2)

esto es una cadena


In [ ]:
"{}{}".format(string_1,string_2)

'esto es una cadena'

In [ ]:
#Armo un string "dinamico", se verá la utilidad cuando trabajemos con ciclos o loops.
string_1 = "dinamica"
string_2 = "esto es una cadena {}".format(string_1)
print(string_2)

esto es una cadena dinamica


In [ ]:
lista_nombres = "Busqueda del valor 2 dentro de una cadena"
#Calculamos e imprimimos el indice en donde arranca el caracter (podemos buscar una palabra también, mostrará la posición en donde comienza dicha palabra)
indice_comienzo = lista_nombres.find("2")
print("la posicion donde se encuentra el 2 es "+str(indice_comienzo))
print("Esta es otra forma de concatenar strings o hacer strings dinamicos con el indice {}".format(str(indice_comienzo)))

#imprimimos la parte de la cadena desde que comienza el caracter 2
print(lista_nombres[indice_comienzo:])

la posicion donde se encuentra el 2 es 19
Esta es otra forma de concatenar strings o hacer strings dinamicos con el indice 19
2 dentro de una cadena


## Ejercicio ##
Concatenar dos cadenas que nos indiquen la palabras "Hello" y "World" (dejar un espacio entre cada palabra)

In [ ]:
string_1 = "hello"
string_2 = " "
string_3 = "world"

print(string_1+string_2+string_3)



string_1 = "hello "
string_2 = "world"
print(string_1+string_2)


string_1 = "hello"
string_2 = " world"

print(string_1+string_2)


string_1 = "hello"
string_2 = "world"


print("{} {}".format(string_1,string_2))



hello world
hello world
hello world
hello world


In [ ]:
lista_nombres = "Busqueda del valor 2 dentro de una cadena"
indice_comienzo = lista_nombres.find("2", 22)
print("no devuelve ningun indice ya que arranca a evaluar a partir de la posicion 22, {}".format(indice_comienzo))
#Veremos que no devuelve nada

no devuelve ningun indice ya que arranca a evaluar a partir de la posicion 22, -1


In [ ]:
#buscar diferentes posiciones dentro de un string
string = "1, 2, 3, 4"
#Primera posicion
print("primera posición en la cadena: {}".format(string[0]))
#Ultima posición
print("ultima posición en la cadena: {}".format(string[-1]))

primera posición en la cadena: 1
ultima posición en la cadena: 4


In [ ]:
string[len(string)-1]
string[-1]

'4'

In [ ]:
#Reemplazar parte de un string 
empresa = "Nuestra empresa es Cordial"
print(empresa)
empresa = empresa.replace("Cordial", "IUDU")
print(empresa)

Nuestra empresa es Cordial
Nuestra empresa es IUDU


In [ ]:
string = "1STO 1S UNA PRUEBA"
string.replace("1", "E")

'ESTO ES UNA PRUEBA'

In [ ]:
string = "1STO 1S UNA PRUEBA"
string[0:string.find("1")+1].replace("1", "E")+string[string.find("1")+1:]


'ESTO 1S UNA PRUEBA'

In [ ]:
string = "1STO 1S UNA PRUEBA"
#string[-6:]

string[4:]



' 1S UNA PRUEBA'

In [ ]:
#Quito espacios de izquierda y de derecha

print("_"*100)

string = "   Se me pegaron muchos espacios    "
print("*"+string+"*")
print("*"+string.strip()+"*")

#sólo quito espacios a derecha
print("*"+string.rstrip()+"*")

#sólo quito espacios a iquierda
print("*"+string.lstrip()+"*")

print("*"*100)


____________________________________________________________________________________________________
*   Se me pegaron muchos espacios    *
*Se me pegaron muchos espacios*
*   Se me pegaron muchos espacios*
*Se me pegaron muchos espacios    *
****************************************************************************************************


In [ ]:
#Convierto todo en mayusculas o en minusculas
string = "Bienvenidos al curso de Machine Learning"
print(string.upper())
print(string.lower())

BIENVENIDOS AL CURSO DE MACHINE LEARNING
bienvenidos al curso de machine learning


In [ ]:
#Convierto todo en "Capital Letter"
string = "bienvenidos al curso de machine learning"
print(string.capitalize())


Bienvenidos al curso de machine learning


In [ ]:
string_1 = "20"
string_2 = "30"

#Concateno las cadenas
print(string_1+string_2)

print("*"*4)
#sumo las cadenas
print(int(string_1)+int(string_2))

2030
****
50


## Ejercicio ##
Teniendo una cadena con el siguiente formato: "5 y 2", restar el primer caracter y el último de la cadena e imprimir el resultado

In [ ]:
string = "5 y 2"

a = int(string[0])-int(string[-1])
print(a)

3


## Operaciones Básicas


| Operador  | Operación                        | Ejemplo         |
| --------- | -------------------------------- | --------------- |
| **        | Potencia                         | `2 ** 3 = 8`    |
| %         | Resto                            | `22 % 8 = 6`    |
| //        | División de enteros              | `22 // 8 = 2`   |
| /         | Division normal                  | `22 / 8 = 2.75` |
| *         | Multiplicación                   | `3 * 3 = 9`     |
| -         | Resta                            | `5 - 2 = 3`     |
| +         | Suma         | `2 + 2 = 4`     |

In [ ]:
operacion = 22/8
print(operacion)

operacion = 22//8
print(operacion)


operacion = 22%8
print(operacion)



2.75
2
6


In [ ]:
operaciones = 51
bach = 4

print(operaciones //4)

operaciones_fuera = operaciones % 4
print(operaciones_fuera)

12
3


## Tipos Básicos de Datos


| Tipo  | Ejemplo
| --------- | --------------- |
| Lógico    | True, False| 
| Entero    | 1, 2, 3| 
| Decimal o punto flotante         | 1.25, -120.25 | 
| Cadena o texto        | Hola, Adiós              | 
| Lista        | \[1, 2, 3\]              |
| Diccionario        | {"llave": "valor"}              |

In [ ]:
#Con esta sentencia vemos todas las variables que están creadas en el entorno
%whos

Variable            Type    Data/Info
-------------------------------------
a                   int     3
bach                int     4
empresa             str     Nuestra empresa es IUDU
indice_comienzo     int     -1
lista_nombres       str     Busqueda del valor 2 dentro de una cadena
operacion           int     6
operaciones         int     51
operaciones_fuera   int     3
string              str     5 y 2
string_1            str     20
string_2            str     30
string_3            str     world


In [ ]:
from datetime import datetime
import time


In [ ]:
datetime.today().year
datetime.today().month
datetime.today().second
datetime.today().hour

10

In [ ]:
a = datetime.today()
time.sleep(10)
print(datetime.today()-a)


0:00:10.001076


In [ ]:
# Loops
for x in range(0,10):
    print("ejecucion numero: {}".format(x))

ejecucion numero: 0
ejecucion numero: 1
ejecucion numero: 2
ejecucion numero: 3
ejecucion numero: 4
ejecucion numero: 5
ejecucion numero: 6
ejecucion numero: 7
ejecucion numero: 8
ejecucion numero: 9


In [ ]:
seguir = True
i = 0
while seguir:
    print("ejecución numero: {}".format(i))
    i+=1
    if i == 10:
        break

ejecución numero: 0
ejecución numero: 1
ejecución numero: 2
ejecución numero: 3
ejecución numero: 4
ejecución numero: 5
ejecución numero: 6
ejecución numero: 7
ejecución numero: 8
ejecución numero: 9


In [ ]:
seguir = True
i = 0
while seguir:
    print("ejecución numero: {}".format(i))
    i+=1
    if i == 10:
        seguir = False

ejecución numero: 0
ejecución numero: 1
ejecución numero: 2
ejecución numero: 3
ejecución numero: 4
ejecución numero: 5
ejecución numero: 6
ejecución numero: 7
ejecución numero: 8
ejecución numero: 9


In [ ]:
# Loops
for x in range(0,100):
    print("*"*20)
    print("ejecucion numero: {}".format(x))
    if x % 2 == 0:
        print("es una ejecucion par")
    elif x == 3:
        pass
        print("Es la ejecucion del número 3")
    else:
        continue
        print("Es la ejecucion del número un numero impar")
    
#Probar cambiando esta linea por if x==11
    if x == 12:
        break

********************
ejecucion numero: 0
es una ejecucion par
********************
ejecucion numero: 1
********************
ejecucion numero: 2
es una ejecucion par
********************
ejecucion numero: 3
Es la ejecucion del número 3
********************
ejecucion numero: 4
es una ejecucion par
********************
ejecucion numero: 5
********************
ejecucion numero: 6
es una ejecucion par
********************
ejecucion numero: 7
********************
ejecucion numero: 8
es una ejecucion par
********************
ejecucion numero: 9
********************
ejecucion numero: 10
es una ejecucion par
********************
ejecucion numero: 11
********************
ejecucion numero: 12
es una ejecucion par


## Ejercicio ##
Sumar los valores del 1 al 100 e imprimir el valor final

In [ ]:
suma = 0
iteraciones = 0
for x in range(0,101):
    suma += x
    iteraciones+=1

print(suma)
print(iteraciones)

suma = 0
iteraciones = 0
for x in range(1,101):
    suma += x
    iteraciones+=1    

print(suma)
print(iteraciones)


if suma == 5050:
    print("hola")

IndentationError: expected an indented block (Temp/ipykernel_2300/454174808.py, line 21)

### Listas ###

Una lista es una estructura de datos que contiene una colección o secuencia de datos. Los datos o elementos de una lista deben ir separados con una coma y todo el conjunto entre corchetes. Se dice que una lista es una estructura mutable porque además de permitir el acceso a los elementos, pueden suprimirse o agregarse nuevos.

lista = ["hola", "mundo", 1, 2, 4.5, ["otra", "lista"]]



In [ ]:
lista = [1, 2, 3, 4]
print("la lista es: {}".format(lista))

la lista es: [1, 2, 3, 4]


In [ ]:
lista.append(5)

In [ ]:
print("la lista actualizada es: {}".format(lista))

la lista actualizada es: [1, 2, 3, 4, 5]


In [ ]:
#Elimino valor por key
lista.pop(-2)

4

In [ ]:
print("la lista actualizada es: {}".format(lista))

la lista actualizada es: [1, 2, 3, 5]


In [ ]:
#Elimino valor por value
lista.remove(5)

In [ ]:
print("la lista actualizada es: {}".format(lista))

la lista actualizada es: [1, 2, 3]


In [ ]:
lista.append(4)
lista.append(5)

In [ ]:
for x in lista:
    print("valor de la lista: {}".format(x))

valor de la lista: 1
valor de la lista: 2
valor de la lista: 3
valor de la lista: 4
valor de la lista: 5


In [ ]:
#otra forma de iterar una lista
lista_nueva = [x*2 for x in lista]

print("la lista nueva es: {}".format(lista_nueva))

la lista nueva es: [2, 4, 6, 8, 10]


In [ ]:
lista_2 = []
for x in lista:
    lista_2.append(x*2)
print(lista_2)

[2, 4, 6, 8, 10]


## Ejercicio ##
De la lista = [2,3,4,5,6] crear una nueva lista con los mismos números potenciados con el mismo numero, e imprimir la nueva lista

Por ejemplo: si el valor es 8, se debería calcular $8^8$

In [ ]:
#itero dos listas en simultaneo
for x, y in zip(lista, lista_nueva):
    print("el valor de la primera lista es: {}, y el valor del a segunda lista es: {}".format(x,y))

el valor de la primera lista es: 1, y el valor del a segunda lista es: 2
el valor de la primera lista es: 2, y el valor del a segunda lista es: 4
el valor de la primera lista es: 3, y el valor del a segunda lista es: 6
el valor de la primera lista es: 4, y el valor del a segunda lista es: 8
el valor de la primera lista es: 5, y el valor del a segunda lista es: 10


In [ ]:
#Modificamos el primer valor de la lista
lista[0] = lista[0]+8
print("la nueva lista modificada es: {}".format(lista))

la nueva lista modificada es: [9, 2, 3, 4, 5]


In [ ]:
print("Longitud de la lista: {} valores".format(len(lista)))

Longitud de la lista: 5 valores


## Tuplas ##

Una tupla permite tener agrupados un conjunto inmutable de elementos, es decir, en una tupla no es posible agregar ni eliminar elementos. Las tuplas se declaran separando los elementos por comas y éstos, opcionalmente, pueden ir entre paréntesis. Se recomienda el uso de paréntesis para evitar ambigüedades del tipo

In [ ]:
tupla = (4,5,6)

In [ ]:
for valor in tupla:
    print("el valor en la tupla es: {}".format(valor))

el valor en la tupla es: 4
el valor en la tupla es: 5
el valor en la tupla es: 6


In [ ]:
#Tirará un error ya que no se puede modificar
tupla[0] = 5

TypeError: 'tuple' object does not support item assignment

In [ ]:
#crear una lista vacía
lista = []
print("imprimo una lista vacia: {}".format(lista))

imprimo una lista vacia: []


In [ ]:
string_1 = "1, 2, 3, 4, 5, 6"
lista_1 = string_1.split(",")
print("lista armada desde un string: {}".format(lista_1))

lista armada desde un string: ['1', ' 2', ' 3', ' 4', ' 5', ' 6']


In [ ]:
#Corrijo la lista sacando los espacios
lista_1 = string_1.replace(" ", "").split(",")
print("lista armada desde un string: {}".format(lista_1))

lista armada desde un string: ['1', '2', '3', '4', '5', '6']


In [ ]:
#Realizo la ultima modificacion para que los valores sean numéricos
lista_1 = string_1.replace(" ", "").split(",")
lista_1 = [int(x) for x in lista_1]
print("lista armada desde un string: {}".format(lista_1))


lista armada desde un string: [1, 2, 3, 4, 5, 6]


## Ejercicio ##
En base a la siguiente lista: string_1 = "1, 2, 3, 4, 5, 6". En una sola línea crear la suma de todos los números e imprimirlo


In [ ]:
#Sumo los valores de la lista
suma_total = 0
suma_total_2 = 0
for valor in lista_1:
    suma_total+=valor
    suma_total_2 = suma_total_2+valor
print("la suma total es: {}".format(suma_total))
print("la suma total es: {}".format(suma_total_2))

la suma total es: 21
la suma total es: 21


In [ ]:
#Otra forma de sumar los valores de una lista
suma_total = sum(lista_1)
print("la suma total es: {}".format(suma_total))
print("la suma total es: {}".format(sum(lista_1)))


la suma total es: 21
la suma total es: 21


In [ ]:
lista_1.sort(reverse=True)
print("lista ordenada descendente: {}".format(lista_1))

lista ordenada descendente: [1, 2, 3, 4, 5, 6]


## Diccionarios ##

Los diccionarios son objetos que contienen una lista de parejas de elementos. De cada pareja un elemento es la clave, que no puede repetirse, y, el otro, un valor asociado. La clave que se utiliza para acceder al valor tiene que ser un dato inmutable como una cadena, mientras que el valor puede ser un número, una cadena, un valor lógico (True/False), una lista o una tupla.

Los pares clave-valor están separados por dos puntos, las parejas por comas y todo el conjunto se encierra entre llaves.

In [ ]:
campeonato_local = {
"River Plate": 1,
"Defensa y Justicia":3,
"Talleres de Cordoba": 2,
"Platense": 19
}

In [ ]:
equipo = "River Plate"
print("la posición que tiene el equipo {} es: {}".format(equipo, campeonato_local[equipo]))

la posición que tiene el equipo River Plate es: 1


In [ ]:
campeonato_local_2 = {}
campeonato_local_2["Julian Alvarez"] = 19
campeonato_local_2["Marco Ruben"] = 16

print("el diccionario de goleadores es: {}".format(campeonato_local_2))


el diccionario de goleadores es: {'Julian Alvarez': 19, 'Marco Ruben': 16}


In [ ]:
campeonato_local_2["Julian Alvarez"]

19

In [ ]:
diccionario = {"v1": 1, "v2": 2, "v3": 3}
for a in diccionario.keys():
    print(a, diccionario[a])

v1 1
v2 2
v3 3


In [ ]:
for a in diccionario:
    print(a, diccionario[a])

v1 1
v2 2
v3 3


In [ ]:
diccionario.items()

AttributeError: 'dict_items' object has no attribute 'values'

In [ ]:
for a, b in diccionario.items():
    print(a,b)

v1 1
v2 2
v3 3


In [ ]:
lista = []
for x in range(100, 501):
    if x%50==0:
        lista.append(x)


In [ ]:
[x for x in range(100,501) if x%50==0]

[100, 150, 200, 250, 300, 350, 400, 450, 500]

## Ejercicios ##

1) Asignar a una variable la división de 450 por 34  
2) De la división anterior, quedarse con el resto de la operación  
3) iterar el número de veces igual al resto obtenido de la operación anterior. En cada una de las iteraciones imprimir el valor e indicar si el mismo es par o impar.  
4) Crear una lista y agregar todos los elementos de la iteración anterior.  
5) Eliminar el valor de la posición 5 y obtener el valor mínimo de la lista  
6) Crear una lista similar a la primera (con los valores del resto) pero todos potenciados a la 2  
7) Imprimo todos los valores (el valor y la potencia) de cada uno de los elementos de las listas  
8) Crear una lista con valores del 100 al 500 a intervalos de 50

Bonus extra 

9) Calculo la raiz cuadrada de 4 
10) Calculo la media de la lista. (Tratar de realizarlo en una sóla línea)